### PR TITLE
[API] As a user, I can Sign-in

### LINKED ISSUE #40: [API] As a user, I can Sign-in
(updating)

## Acceptance criteria
(updating)

### PR DESCRIPTION
**Pull request by byhbt**
Issue: https://github.com/byhbt/kwtool/issues/40

## What happened

As a user, I can Sign-in.
 
## Insight

- [x] Implement `error_view` for API.
- [x] Implement Login API.

API docs:
- https://documenter.getpostman.com/view/104198/UVXkovCX

## Proof Of Work

The tests should pass.

https://user-images.githubusercontent.com/948856/150103780-9e3b0a00-6844-4b7b-9070-112250519f9a.mp4

### CODE REVIEW

File: lib/kwtool_web/controllers/api/v1/session_controller.ex [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  alias Kwtool.Account.{Guardian, Users}
+  alias KwtoolWeb.Api.V1.ErrorView
+
+  def create(conn, %{"email" => email, "password" => password}) do
+    case Users.authenticate_user(email, password) do
+      {:ok, user} ->
+        {:ok, jwt, _full_claims} = Guardian.encode_and_sign(user, %{})
+
+        render(conn, "show.json", %{data: %{id: user.id, email: user.email, jwt: jwt}})
```
Comment (rosle) [inline lib/kwtool_web/controllers/api/v1/session_controller.ex]: For the ID of login resource, maybe we can generate some value e.g. timestamps.

File: test/kwtool_web/views/api/v1/error_view_test.exs [hunk 2]
```diff
@@ -0,0 +1,96 @@
+defmodule KwtoolWeb.Api.V1.ErrorViewTest do
+  use KwtoolWeb.ConnCase, async: true
+
+  # Bring render/3 and render_to_string/3 for testing custom views
+  import Phoenix.View
```
Comment (rosle) [inline test/kwtool_web/views/api/v1/error_view_test.exs]: Maybe we can use ViewCase instead

File: lib/kwtool_web/router.ex [hunk 3]
```diff
@@ -23,6 +23,19 @@ defmodule KwtoolWeb.Router do
...
+      :api,
+      KwtoolWeb.CheckEmptyBodyParamsPlug
+    ]
+  end
+
+  scope "/api/v1", KwtoolWeb.Api.V1, as: :api do
+    pipe_through :api
+
+    post "/sign_in", SessionController, :create
+  end
```
Comment (rosle) [inline lib/kwtool_web/router.ex]: Does this work? 🤔 💡 

```suggestion
  scope "/api/v1", KwtoolWeb.Api.V1, as: :api do
    pipe_through [
      :api,
      KwtoolWeb.CheckEmptyBodyParamsPlug
    ]

    post "/sign_in", SessionController, :create
  end
```

File: test/kwtool_web/controllers/api/v1/error_handler_test.exs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+defmodule KwtoolWeb.Api.V1.ErrorHandlerTest do
+  use KwtoolWeb.ConnCase, async: true
+
+  alias KwtoolWeb.Api.V1.ErrorHandler
+
+  describe "auth_error/2" do
+    test "renders 401 error page", %{conn: conn} do
```
Comment (hanam1ni) [inline test/kwtool_web/controllers/api/v1/error_handler_test.exs]:
```suggestion
    test "renders 401 error response", %{conn: conn} do
```

File: test/kwtool_web/views/api/v1/error_view_test.exs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      ])
+      |> validate_required([
+        :device_id,
+        :operating_system,
+        :device_name
+      ])
+    end
+  end
+
+  test "renders 404.json" do
```
Comment (hanam1ni) [inline test/kwtool_web/views/api/v1/error_view_test.exs]: Let's put all the test cases here inside `describe` block

File: test/kwtool_web/views/api/v1/error_view_test.exs (hunk 2 above)
Comment (byhbt) [reply test/kwtool_web/views/api/v1/error_view_test.exs]: Yes, i updated in https://github.com/byhbt/kwtool/pull/45/commits/305904ac02be117c7e1f55d5ef9479984d526519

File: lib/kwtool_web/controllers/api/v1/session_controller.ex (hunk 1 above)
Comment (byhbt) [reply lib/kwtool_web/controllers/api/v1/session_controller.ex]: Updated in https://github.com/byhbt/kwtool/pull/45/commits/d4700d1b319c67aa78cd2b7a4e64e1e10af3378a

File: lib/kwtool_web/router.ex (hunk 3 above)
Comment (byhbt) [reply lib/kwtool_web/router.ex]: Updated in https://github.com/byhbt/kwtool/pull/45/commits/5c56d403ac73dd8c41a089f0f6397895627d3386

File: test/kwtool_web/views/api/v1/error_view_test.exs (hunk 5 above)
Comment (byhbt) [reply test/kwtool_web/views/api/v1/error_view_test.exs]: Wrapped in https://github.com/byhbt/kwtool/pull/45/commits/9c44d18b14f6ef16749063289d5ec0cab7b29835

File: test/kwtool_web/controllers/api/v1/error_handler_test.exs (hunk 4 above)
Comment (byhbt) [reply test/kwtool_web/controllers/api/v1/error_handler_test.exs]: Updated in https://github.com/byhbt/kwtool/pull/45/commits/5c56d403ac73dd8c41a089f0f6397895627d3386

**Review by hanam1ni — APPROVED** [review]

**Review by rosle — APPROVED** [review]

**Pull request merged**